gflownet/envs/crystals/composition.py

```diff
@@ -215,12 +228,21 @@ def get_mask_invalid_actions_forward(self, state=None, done=None):
             min_atoms_per_unrequired_element = 0
             max_atoms_per_unrequired_element = 0
 
+        # Check if current state is compatible with space group
+        if self.do_spacegroup_check and isinstance(self.space_group, int):
+            n_atoms = [s for s in state if s > 0]
+            sg_compatible = all(Group(self.space_group).check_compatible(n_atoms))
```

Comment:
It's fast enough, I think. I have the statistics indeed. Below are the mean, standard deviation, minimum and maximum time over the mean times of each space group.

- mean : 0.62 ms
- standard deviation: 0.62 ms
- minimum: 0.05 ms
- maximum: 3.81 ms

Reply:
Ok, that seems fast enough. 

For the Dav(e) proxy, I think it would be around 2-3 seconds per iteration for a batch size of 32, with the improvements currently in main.
That means 32 * 0.62 ms  = about 20ms / iteration a.k.a. it would add about 1% to the runtime.

Not an issue at all, I think.

Reply:
I forgot to account for the fact that this would be called every time the mask is computed so if there is an average of 4 composition actions per trajectory, that would be 4% of the runtime. Still not an issue, I think, but I wanted to be thorough.

Reply:
(sorry, unresolving because I saw something worth discussing)
On my branch, I was testing to see how much time different alternative ways to filter actions in the mask (based on space group) could take. I was using space group 227 as a first test bed and the timings I saw were usually around 40ms on average with the max around 200ms.
We probably can't directly compare the two averages because I was testing on a much more restrictive space group but I'm still baffled that I see a maximum about 40x larger than what you saw.

Reply:
The numbers I gave are means of means :sweat_smile: which could explain the difference. To clarify:
- I initially computed the mean time over all compositions tested for each space group (see [this script](https://github.com/alexhernandezgarcia/gflownet/blob/composition-sg-constraints/scripts/pyxtal/get_n_compatible_for_sg.py))
- I reported above the mean of all those means (over all space groups), and std, min, max. 

That said, specificaly for space group 227 I got a much lower mean time (0.0003 s = 0.3 ms):

```
space_group,n_compatible,n_not_compatible,n_total,time_mean,time_std
227,151,4083950,4084101,0.00030409366723782566,7.457020201488068e-05
```

Reply:
I did it in a compute node of the cluster, by the way.

Reply:
Finally figure out the difference. Turns out that instantiating a PyXtal space group is really expensive (like... _really_ expensive).
In your script, you instantiate once and reuse. In my timings, I was re-instantiating each time so that was much more expensive. When I instantiate once and reuse, I get the same timings as you do.

Reply:
I see! Good finding and very good to know!